ledger/streaming/client/src/scp_validator.rs

```diff
@@ -0,0 +1,362 @@
+    }
+}
+
+impl<US: BlockStream + 'static, ID: GenericNodeId + Send> BlockStream for SCPValidator<US, ID> {
+    type Stream = impl Stream<Item = StreamResult<BlockStreamComponents>>;
+
+    /// Get block stream that performs validation
+    fn get_block_stream(&self, starting_height: u64) -> StreamResult<Self::Stream> {
+        // Merge all streams into one
+        let mut merged_streams = stream::SelectAll::new();
+        for stream_factory in &self.upstreams {
+            let (node_id, upstream) = stream_factory;
+            log::info!(self.logger, "Generating new stream from {:?}", node_id);
+            let us = upstream.get_block_stream(starting_height).unwrap();
+            let peer_id = node_id.clone();
+            merged_streams.push(Box::pin(
+                us.map(move |component| (peer_id.clone(), component)),
+            ));
+        }
+
+        // Create SCP validation state object and insert it into stateful stream
+        let mut validation_state = self.scp_validation_state.clone();
+        if starting_height > 0 {
+            validation_state.set_highest_externalized_slot(starting_height - 1);
+        }
+
+        Ok(merged_streams
+            .scan(validation_state, |scp_state, (node_id, component)| {
+                if let Ok(component) = component {
+                    // Put component into underlying state
+                    scp_state.update_slot(node_id, component);
+
+                    // Attempt to externalize it if there's quorum
+                    if let Some(component) = scp_state.attempt_externalize_block() {
+                        return future::ready(Some(Ready::Ready(Ok(component))));
+                    }
+                }
+
+                if scp_state.is_behind() {
+                    return future::ready(Some(Ready::Ready(Err(StreamError::ConsensusBlocked(
+                        scp_state.highest_externalized_slot(),
+                        scp_state.highest_block_received(),
+                    )))));
+                }
+                // If there's not quorum yet, forward NotReady to filter stream
+                future::ready(Some(Ready::NotReady))
+            })
+            .filter_map(|result| {
+                // If the component is in ready state forward it to next stream
+                if let Ready::Ready(component) = result {
+                    return future::ready(Some(component));
+                }
+
+                // If no ready message was received don't send anything onwards
+                // (FilterMap streams don't forward None values)
+                future::ready(None)
+            }))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use mc_common::logger::{test_with_logger, Logger};
+    use mc_consensus_scp::test_utils::test_node_id;
+    use mc_ledger_streaming_api::test_utils::stream::SimpleMockStream;
+    use mc_transaction_core::BlockIndex;
+
+    use super::*;
+
+    #[test_with_logger]
+    fn scp_validates_nodes_in_quorum(logger: Logger) {
```

Comment:
Please add more tests exercising some cases where we cannot reach quorum, e.g. too few votes, conflicting votes. It also seems useful to exercise some of the validation logic, e.g. `threshold > members.len()`, `threshold = 0`, etc.